lib/utils/archive.go

```diff
@@ -75,3 +82,83 @@ func CompressTarGzArchive(files []string, fileReader ReadStatFS) (*bytes.Buffer,
 
 	return archiveBytes, nil
 }
+
+// CompressDirToZipFile compresses directory into a `.zip` format.
+func CompressDirToZipFile(sourcePath, destPath string) error {
+	archive, err := os.Create(destPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	zipWriter := zip.NewWriter(archive)
+	err = filepath.Walk(sourcePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		zipFileWriter, err := zipWriter.Create(filepath.Base(path))
+		if err != nil {
+			return trace.NewAggregate(err, file.Close())
+		}
+
+		_, err = io.Copy(zipFileWriter, file)
+		return trace.NewAggregate(err, file.Close())
+	})
+
+	return trace.NewAggregate(err, zipWriter.Close(), archive.Close())
+}
+
+// CompressDirToTarGzFile compresses directory into a `.tar.gz` format.
+func CompressDirToTarGzFile(sourcePath, destPath string) error {
+	archive, err := os.Create(destPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	gzipWriter := gzip.NewWriter(archive)
+	tarWriter := tar.NewWriter(gzipWriter)
+	err = filepath.Walk(sourcePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		header, err := tar.FileInfoHeader(info, info.Name())
+		if err != nil {
+			return trace.NewAggregate(err, file.Close())
+		}
+		if err := tarWriter.WriteHeader(header); err != nil {
+			return trace.NewAggregate(err, file.Close())
+		}
+
+		_, err = io.Copy(tarWriter, file)
+		return trace.NewAggregate(err, file.Close())
+	})
+
+	return trace.NewAggregate(err, tarWriter.Close(), gzipWriter.Close(), archive.Close())
```

Comment:
There are a few ways to avoid the need to repeatedly close on return. One "magical" way is to name the return:

```golang
func doThing() (err error) {
	f, err := os.Create(...)
	...
	defer func() {
		if cErr := f.Close(); err == nil {
			err = cErr
		} 
        }()
}
```
An advantage of this approach is that it won't leak FDs on panic, but I generally avoid it because it's difficult to understand.

In cases where you have multiple wrapped writers that need to be closed, I'd suggest breaking out a helper method:
```golang
func CompressDirToTarGzFile(sourcePath, destPath string) error {
	archive, err := os.Create(destPath)
	if err != nil {
		return trace.Wrap(err)
	}
	gzipWriter := gzip.NewWriter(archive)
	tarWriter := tar.NewWriter(gzipWriter)
        err := doWalk(tarWriter)
        if err != nil {
                _ = archive.Close()
                return trace.Wrap(err)
        }
        if err := tarWriter.Close(); err != nil {
                _ = archive.Close()
                return trace.Wrap(err)
        }
        if err := gzipWriter.Close(); err != nil {
                _ = archive.Close()
                return trace.Wrap(err)
        }
        return trace.Wrap(archive.Close())
}
```
(Personally I avoid the aggregation unless the errors are all important.)

Reply:
Actually, it looks like they've defined ErrClosed for os.File now, so it's safe to call multiple times:
```golang
func CompressDirToTarGzFile(sourcePath, destPath string) error {
	archive, err := os.Create(destPath)
	if err != nil {
		return trace.Wrap(err)
	}
	defer archive.Close()
	gzipWriter := gzip.NewWriter(archive)
	tarWriter := tar.NewWriter(gzipWriter)
        err := doWalk(tarWriter)
        if err != nil {
                return trace.Wrap(err)
        }
        if err := tarWriter.Close(); err != nil {
                return trace.Wrap(err)
        }
        if err := gzipWriter.Close(); err != nil {
                return trace.Wrap(err)
        }
        return trace.Wrap(archive.Close())
}
```

Reply:
Context: https://github.com/golang/go/issues/32427